src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1072 @@
+
+                except Exception:
+                    self.health = "{} Exception at {}".format(
+                        sys.exc_info()[0].__name__,
+                        datetime.strftime(datetime.now(),"%Y/%m/%d %H:%M:%S")
+                    )
+                    log.exception(self.health)
+                    self.active = False
+                    break
+
+            log.warning("[WRN] Namespace event watcher stopped")
+
+
+class KubernetesEvent(RookCeph):
+
+    def __init__(self, log_entry, unique_name=True):
+        super(KubernetesEvent, self).__init__()
+
+        self.event_name = log_entry.event_name
+        self.message = log_entry.event_msg
+        self.event_type = log_entry.event_type
+        self.event_reason = log_entry.event_reason
+        self.unique_name = unique_name
+
+        self.api_status = 200
+        self.count = 1
+        self.first_timestamp = None
+        self.last_timestamp = None
+
+    @property
+    def event_body(self):
+        if self.unique_name:
+            obj_meta = client.V1ObjectMeta(name="{}".format(self.event_name)) 
+        else:
+            obj_meta = client.V1ObjectMeta(generate_name="{}".format(self.event_name))
+
+        # field_path is needed to prevent problems in the namespacewatcher when
+        # deleted event are received
+        obj_ref = client.V1ObjectReference(kind="CephCluster",
+                                           field_path='spec.containers{mgr}', 
+                                           name=self.event_name, 
+                                           namespace=self.namespace)
+
+        event_source = client.V1EventSource(component="ceph-mgr", 
+                                            host=self.host)
+        return  client.V1Event(
+                    involved_object=obj_ref, 
+                    metadata=obj_meta, 
+                    message=self.message, 
+                    count=self.count, 
+                    type=self.event_type,
+                    reason=self.event_reason,
+                    source=event_source, 
+                    first_timestamp=self.first_timestamp,
+                    last_timestamp=self.last_timestamp
+                )
+
+    def write(self):
+
+        now=datetime.now(timezone.utc)
+
+        self.first_timestamp = now
+        self.last_timestamp = now
+
+        try:
+            self.api.create_namespaced_event(self.namespace, self.event_body)
+        except MaxRetryError:
+            # k8s config has not be defined properly 
+            self.api_status = 403       # Forbidden
+        except ApiException as e:
+            self.api_status = e.status
+            if e.status == 409:
+                # 409 means the event is there already, so read it back (v1Event object returned)
+                # this could happen if the event has been created, and then the k8sevent module
+                # disabled and reenabled - i.e. the internal event tracking no longer matches k8s
```

Comment:
This will also happen every time the mgr pod is restarted, right? The event cache is just in memory. Should we be more proactive about rehydrating the events from K8s when the module starts up?

Reply:
there are two categories of events - those we send to k8s and those we fetch from k8s. The k8s events we fetch are always in sync as long as the namespacewatcher thread is alive. The events we send however would start from scratch following a mgr restart, so yes you'd lose history. However, events have a TTL in k8s anyway so they're not meant for a long term history of what's gone on. Our events cache goes beyond normal event retention - if we want to provide a history of the events we've sent to k8s we'll need to persist them to the monstore. I just didn't feel that that was warranted given the normal event lifecycle in k8s.

Reply:
Right, and as discussed we can start here and consider persistence for the events if it ends up being more common than expected.